### PR TITLE
chore: complete apdex automation tasks

### PR DESCRIPTION
**Pull request by ralfudx**
# Description

Merge [test scenario to capture telemetry data for tasks:refresh](https://github.com/medic/care-teams/issues/135) into apdex-automation-tests branch and clean up code

[#142](https://github.com/medic/care-teams/issues/142)

# Code review checklist

- [ ] Readable: Concise, well named, follows the [style guide](https://docs.communityhealthtoolkit.org/contribute/code/style-guide/), documented if necessary.
- [ ] Documented: Configuration and user documentation on [cht-docs](https://github.com/medic/cht-docs/)
- [ ] Tested: Unit and/or e2e where appropriate
- [ ] Internationalised: All user facing text
- [ ] Backwards compatible: Works with existing data and configuration or includes a migration. Any breaking changes documented in the release notes.

# Compose URLs
If Build CI hasn't passed, these may 404:

* [Core](https://staging.dev.medicmobile.org/_couch/builds_4/medic:medic:complete-apdex-automation-tasks/docker-compose/cht-core.yml)
* [CouchDB Single](https://staging.dev.medicmobile.org/_couch/builds_4/medic:medic:complete-apdex-automation-tasks/docker-compose/cht-couchdb.yml)
* [CouchDB Cluster](https://staging.dev.medicmobile.org/_couch/builds_4/medic:medic:complete-apdex-automation-tasks/docker-compose/cht-couchdb-clustered.yml)

# License

The software is provided under AGPL-3.0. Contributions to this project are accepted under the same license.

### CODE REVIEW

**Comment by ralfudx** [comment]
Hey the build finally passed thanks to [Diana's contribution](https://medic.slack.com/archives/C024KTGRW/p1720624580645419?thread_ts=1720616206.298739&cid=C024KTGRW) - @latin-panda @tatilepizs kindly have a look when you have some time

File: tests/performance/apdex-score/test/page-objects/contacts.page.js [hunk 1]
```diff
@@ -5,6 +5,7 @@ class ContactsPage extends Page {
   async loadContactList(settingsProvider) {
     const page = settingsProvider.getPage('contact-list');
     await super.loadAndAssertPage(page);
+    await super.toggleAirplaneMode('on');
```
Comment (latin-panda) [inline tests/performance/apdex-score/test/page-objects/contacts.page.js:8]: Is the idea to turn on the airplane mode after login? 
If yes, it's better to make a separate function like this:

```js
// In load.page.js
asyn turnOnAirplaneMode(settingsProvider) {
  const commonElements = settingsProvider.getCommonElements();
  const UI_ELEMENT = commonElements?.relaunchAppAssert || '//*[@text="People"]';
  await this.waitForDisplayedAndRetry(UI_ELEMENT);
  await super.toggleAirplaneMode('on');
}
```
You do something similar for relaunching the app.
Then call that new function after the login in `test.apdex.js`

File: tests/performance/apdex-score/test/page-objects/page.js [hunk 2]
```diff
@@ -176,86 +176,31 @@ module.exports = class Page {
...
-  }
-
-  get menuItemAbout() {
-    return $('//android.view.MenuItem[@text="About"]');
-  }
-
-  get menuItemSettings() {
-    return $('//android.view.MenuItem[@text="User settings"]');
-  }
+  async fillUpFormOnly(form, commonElements) {
```
Comment (latin-panda) [inline tests/performance/apdex-score/test/page-objects/page.js:179]: You're adding this new function because tasks don't come from a FAB button, right?

In that case, to avoid repeating code add a new setting option in the forms like:
```js
 "forms": {
     "patientTask": {
           "navigation": [.......],
           "skipFAB": true, <-----This 
           "pages": [.......]
```
Then update the README.md in [the form's section to describe that new option ](https://github.com/medic/cht-core/tree/complete-apdex-automation-tasks/tests/performance/apdex-score#forms)

Lastly, go the [fillUpForm function](https://github.com/medic/cht-core/blob/cb9eb8b2072f73e382dd11b7ebf6dddf5b50f158/tests/performance/apdex-score/test/page-objects/page.js#L142) and add a IF statement for that

```js
 async fillUpForm(form, commonElements){
    const FAB_SELECTOR = commonElements?.fab || '//android.widget.Button[not(@text="Actions menu")]';
    const FAB_LIST_TITLE = commonElements?.fabListTitle || '//android.widget.TextView[@text="New"]';
    const FORM_SUBMIT_SELECTOR = commonElements?.formSubmit || '//android.widget.Button[@text="Submit"]';
    const FORM_PAGE_NEXT_SELECTOR = commonElements?.formNext || '//android.widget.Button[@text="Next >"]';

   if (form.skipFAB) {
      await this.clickElement(FAB_SELECTOR);
      await this.waitForDisplayedAndRetry(FAB_LIST_TITLE);
   } 
    await this.navigate(form.navigation);

**Review by latin-panda — CHANGES_REQUESTED** [review]
Looking good
Please do another check to remove all unused code and unused files, just to leave this nice and tidy

File: tests/performance/apdex-score/test/page-objects/contacts.page.js (hunk 1 above)
Comment (ralfudx) [reply tests/performance/apdex-score/test/page-objects/contacts.page.js:8]: Great thanks

File: tests/performance/apdex-score/test/page-objects/load.page.js [hunk 3]
```diff
@@ -1,19 +1,23 @@
 const { $ } = require('@wdio/globals');
 const Page = require('./page');
 
-class LoadPage extends Page {
-   
-  get inputInstanceUrl () {
+class LoadPage extends Page {  
+  
+  get btnCustom() {
+    return $('//*[@text="Custom"]');
```
Comment (tatilepizs) [inline tests/performance/apdex-score/test/page-objects/load.page.js:7]: It is better to specify the type of element instead of using `*` I guess that might be `android.widget.Button`, not sure, can you please check?

File: tests/performance/apdex-score/test/page-objects/login.page.js [hunk 4]
```diff
@@ -3,23 +3,23 @@ const Page = require('./page');
...
   
-  get inputUsername () {
+  get inputUsername() {
     return $('//android.view.View[1]/android.widget.EditText');
   }
 
-  get inputPassword () {
+  get inputPassword() {
     return $('//android.view.View[3]/android.widget.EditText');
   }
```
Comment (tatilepizs) [inline tests/performance/apdex-score/test/page-objects/login.page.js:12]: I think that creating a selector referring the element number might cause problems in the future. Please check if you can use text, for example, to access the element in a better way.

Also, maybe we can create a method in `page.js` that returns `inputs`, for example, and you only need to send as a parameter the text of the input that you need. Something similar to what we did on the [common-enketo.wdio.page.js ](https://github.com/medic/cht-core/blob/master/tests/page-objects/default/enketo/common-enketo.wdio.page.js) file

File: tests/performance/apdex-score/test/page-objects/login.page.js [hunk 5]
```diff
@@ -3,23 +3,23 @@ const Page = require('./page');
...
 
-  get btnLogin () {
+  get btnLogin() {
     return $('//*[@text="Login"]');
   }
 
-  get btnAccept () {
+  get btnAccept() {
     return $('//*[@text="Accept"]');
   }
```
Comment (tatilepizs) [inline tests/performance/apdex-score/test/page-objects/login.page.js:20]: Same as the previous comment, maybe we can have a common `get button` method in `page.js` and you only need to send the text of the button that you want to access as a parameter.

**Review by tatilepizs — CHANGES_REQUESTED** [review]
Thank you Rafa for updating the tests ⭐ 

I left a couple of suggestions, and have one question. 

Maybe I am a little late on this, but is there a specific reason why you decided to add a `page-objects` folder inside the `apdex-score/test` folder instead of adding a `apdex` folder inside `the page-objects` folders that we already have?

I think it will be better to have all the page objects files in one place as we are doing with all the other tests suites.

**Comment by ralfudx** [comment]
Thanks for the comments and suggestions @tatilepizs I'll have a look and see if the recommendations can fit in
Also, I guess we can have a quick chat over this during one of the QA sync calls to better align on some of these suggestions

**Comment by ralfudx** [comment]
@tatilepizs In order not to delay the [apdex tasks completion work](https://github.com/medic/care-teams/issues/142) I'd suggest we include these structural changes in a different/subsequent PR
This will allow us to quickly merge this [PR](https://github.com/medic/cht-core/pull/9247) to the apdex-automation-tests branch (as part of this week's commitment for care team) so that @latin-panda is unblocked with her additional changes for the settings file

**Comment by tatilepizs** [comment]
Ok, that would be fine. but can you please create a ticket to address those changes? if we don't do it we will probably will forget and those e2e tests will start growing and then it will be harder to make those changes.

**Comment by ralfudx** [comment]
Sure [this](https://github.com/medic/care-teams/issues/145) has been created to track

**Comment by latin-panda** [comment]
>  I'd suggest we include these structural changes in a different/subsequent PR

@ralfudx Do you mean this feedback ⬇️ from Tatiana? 

> Maybe I am a little late on this, but is there a specific reason why you decided to add a page-objects folder inside the apdex-score/test folder instead of adding a apdex folder inside the page-objects folders that we already have?


If it's just moving files around to another folder and updating imports, let's do it! Because it's not a big coding change.

@tatilepizs @ralfudx do you mean something like this? ⬇️

|-tests/
|---performance/
|------apdex-score/
|---------page-objects/
|------------page.js
|---------tests.js <--- a file containing the test cases (currently apdex-score/test/specs/num-1/test.apdex.js)

@ralfudx please remember to remove all unused files and code. For example we don't need anymore these files because we use setting files to test different deployments:
Screenshot 2024-07-16 at 1 14 54 PM

That leaves us with 1 file only with test cases: `apdex-score/test/specs/num-1/test.apdex.js`. So, simplifying the folders makes sense because we only have 1 file, so following the suggestion from restructuring, it's something like this: `apdex-score/tests.js`.

**Comment by ralfudx** [comment]
@latin-panda I have removed the unused test files

I believe this is the structure @tatilepizs is referring to:
|-tests/
|---page-objects/
|------apdex <---- a folder containing all the apdex page.js files (moved from apdex-score/test/page-objects)
|---performance/
|------apdex-score/
|---------specs
|------------test.apdex.js <--- a file containing the test cases (previously apdex-score/test/specs/num-1/test.apdex.js)

Other suggestions include code changes related to [selector definition](https://github.com/medic/cht-core/pull/9247#discussion_r1674375902), using a [method to capture selectors](https://github.com/medic/cht-core/pull/9247#discussion_r1674384723) and also changes to the wdio.conf.js file similar to what was done in [cht-form](https://github.com/medic/cht-core/blob/complete-apdex-automation-tasks/tests/integration/cht-form/wdio.conf.js)

If we intend to do all these changes in this PR I cannot commit to completing this work Tuesday EOD as mentioned in [Care Team's Chat](https://medic.slack.com/archives/C02GME3BQE6/p1721060268713109)
Also, I believe these changes are out-of-scope for the work intended here and should/can be captured in the larger [apdex-automation-tests Draft PR](https://github.com/medic/cht-core/pull/9037)

We also need to keep in mind that in the future, we may use this project for other types of mobile testing apart from apdex

File: tests/performance/apdex-score/test/page-objects/login.page.js (hunk 5 above)
Comment (latin-panda) [reply tests/performance/apdex-score/test/page-objects/login.page.js:20]: @ralfudx this one is fast, you do it like this: 

```
get button(label) {
    return $('//*[@text="' + label +'"]');
}
```

Then you use it like this:

```
loginPage.button('Login');
loginPage.button('Accept'); 
```

**Comment by latin-panda** [comment]
> I believe this is the structure @tatilepizs is referring to:
|-tests/
|---page-objects/
|------apdex <---- a folder containing all the apdex page.js files (moved from apdex-score/test/page-objects)
|---performance/
|------apdex-score/
|---------specs
|------------test.apdex.js <--- a file containing the test cases (previously apdex-score/test/specs/num-1/test.apdex.js)

@ralfudx, that structure still makes sense to me 👍

[I sent you a snippet of the button code here, small & safe change](https://github.com/medic/cht-core/pull/9247#discussion_r1678978110). I didn't have time to help and test the XPATH selector's feedback. 

> If we intend to do all these changes in this PR I cannot commit to completing this work Tuesday EOD as mentioned in [Care Team's Chat](https://medic.slack.com/archives/C02GME3BQE6/p1721060268713109)
Also, I believe these changes are out-of-scope for the work intended here and should/can be captured in the larger https://github.com/medic/cht-core/pull/9037

It seems Michael is requesting the entire [Apdex work](https://github.com/medic/cht-core/pull/9037) to be merged directly into cht-core by EOD Tuesday, as mentioned in the [Care Teams' summary](https://medic.slack.com/archives/C02GME3BQE6/p1721060268713109). The intention is to effectively complete this feature that [has been open since April ](https://github.com/medic/cht-core/pull/9037). Please don't compromise code and solution quality, and do not let tickets be done later because this is still a manageable size of project and historically we don't retake those tickets in years. 
To summarize, it's fine to do some of these fixes in the [larger PR](https://github.com/medic/cht-core/pull/9037), but not leave them for later.

File: tests/performance/apdex-score/test/page-objects/load.page.js (hunk 3 above)
Comment (ralfudx) [reply tests/performance/apdex-score/test/page-objects/load.page.js:7]: it's `android.widget.TextView` thanks

File: tests/performance/apdex-score/test/page-objects/login.page.js (hunk 4 above)
Comment (ralfudx) [reply tests/performance/apdex-score/test/page-objects/login.page.js:12]: Done thanks

File: tests/performance/apdex-score/test/page-objects/login.page.js (hunk 5 above)
Comment (ralfudx) [reply tests/performance/apdex-score/test/page-objects/login.page.js:20]: `A 'get' accessor cannot have parameters.ts(1054)`

**Comment by ralfudx** [comment]
@latin-panda @tatilepizs kindly re-review
Thanks

File: tests/page-objects/apdex/page.js [hunk 6]
```diff
@@ -155,17 +157,20 @@ module.exports = class Page {
 
     await this.clickElement(FORM_SUBMIT_SELECTOR);
     await this.assertMany(form.postSubmitAsserts);
-    await this.navigate(form.postTestPath);
+    if (form.useFAB) {
+      await this.navigate(form.postTestPath);
+    }
```
Comment (latin-panda) [inline tests/page-objects/apdex/page.js:162]: I think this shouldn't be depending on FAB. if a form doesn't have a postTestPath, just don't execute it. 
```suggestion
   await this.navigate(form.postTestPath);
```

File: tests/page-objects/apdex/page.js [hunk 7]
```diff
@@ -268,60 +199,8 @@ module.exports = class Page {
...
-    console.log(`TIME IS::: ${await driver.getDeviceTime()}`);
-    //change date here - WIP
-    await browser.pause(5000);
-    await this.clickDisplayedElem(this.tabDropdown);
-    await this.clickDisplayedElem(this.menuItemSyncNow);
-    await browser.pause(1000);
-    await this.tabDropdown.click();
+  async getButton(label) {
+    return $('//*[@text="' + label +'"]');
   }
```
Comment (latin-panda) [inline tests/page-objects/apdex/page.js:204]: Because not using `await` not need of `async` 
```suggestion
 getButton(label) {
    return $('//*[@text="' + label +'"]');
  }
```

File: tests/performance/apdex-score/README.md [hunk 8]
```diff
@@ -8,7 +8,7 @@
 
 <details> <summary>Expand to see settings file structure </summary>
 
-```
+```json {"id":"01J2WE5FDM112EVC71G6QBXS1N"}
```
Comment (latin-panda) [inline tests/performance/apdex-score/README.md:11]: Not sure what is that, but does it work without it?
```suggestion
```json
```

**Review by latin-panda — COMMENTED** [review]
Thanks for cleaning all that code! 🤩
I just did a quick pass because I'm running out of time, I won't be able to review for the rest of the week, so I'm unblocking this PR and let Tatiana continue with the review.

File: tests/page-objects/apdex/page.js (hunk 6 above)
Comment (ralfudx) [reply tests/page-objects/apdex/page.js:162]: Done thanks

File: tests/page-objects/apdex/load.page.js [hunk 9]
```diff
@@ -1,23 +1,30 @@
 const { $ } = require('@wdio/globals');
```
Comment (tatilepizs) [inline tests/page-objects/apdex/load.page.js:1]: Is this necessary? Can you please check if the test will pass if you remove this line?

**Review by tatilepizs — CHANGES_REQUESTED** [review]
Thank you for all the changes @ralfudx ! ⭐ 
I left some small changes

File: tests/page-objects/apdex/tasks.page.js [hunk 10]
```diff
@@ -0,0 +1,20 @@
+const Page = require('./page');
```
[the remark this answers is not in the document]
Comment (tatilepizs) [reply tests/page-objects/apdex/tasks.page.js:1]: Because it is easier to understand and track when anyone reads the code. Also, it is about structure and organization, which is what we have been trying to improve in `/test` folder.

Here are some of the files that are inside `page-object` folder and reference the other files using `@page-objects`:

- [death-report.page.js](https://github.com/medic/cht-core/blob/master/tests/page-objects/default/enketo/death-report.page.js#L3)
- [contacts.wdio.page.js](https://github.com/medic/cht-core/blob/master/tests/page-objects/default/contacts/contacts.wdio.page.js#L1-L8)
- [search.wdio.page.js](https://github.com/medic/cht-core/blob/master/tests/page-objects/default-mobile/search/search.wdio.page.js#L1)
- [common.wdio.page.js](https://github.com/medic/cht-core/blob/master/tests/page-objects/default/common/common.wdio.page.js#L3)
- [target-aggregates.wdio.page.js](https://github.com/medic/cht-core/blob/master/tests/page-objects/default/targets/target-aggregates.wdio.page.js#L1)

The config of `@page-objects` is in the file [`tests/jsconfig.json`](https://github.com/medic/cht-core/blob/master/tests/jsconfig.json#L8)

File: tests/page-objects/apdex/load.page.js (hunk 9 above)
Comment (ralfudx) [reply tests/page-objects/apdex/load.page.js]: Removed thanks

File: tests/page-objects/apdex/login.page.js [hunk 11]
```diff
@@ -0,0 +1,26 @@
+const { $ } = require('@wdio/globals');
+const Page = require('./page');
```
[the remark this answers is not in the document]
Comment (ralfudx) [reply tests/page-objects/apdex/login.page.js]: done thanks

File: tests/performance/apdex-score/specs/test.apdex.js [hunk 12]
```diff
@@ -1,12 +1,12 @@
...
-const tasksPage = require('../../page-objects/tasks.page');
-const messagesPage = require('../../page-objects/messages.page');
-const reportsPage = require('../../page-objects/reports.page');
+const loadPage = require('../../../page-objects/apdex/load.page');
+const loginPage = require('../../../page-objects/apdex/login.page');
+const contactsPage = require('../../../page-objects/apdex/contacts.page');
+const performancePage = require('../../../page-objects/apdex/performance.page');
+const tasksPage = require('../../../page-objects/apdex/tasks.page');
+const messagesPage = require('../../../page-objects/apdex/messages.page');
+const reportsPage = require('../../../page-objects/apdex/reports.page');
```
[the remark this answers is not in the document]
Comment (ralfudx) [reply tests/performance/apdex-score/specs/test.apdex.js]: Done thanks

**Comment by ralfudx** [comment]
@tatilepizs all requested changes have now been resolved - kindly have take out some time to review again on Monday thanks for all the feedback

**Review by tatilepizs — APPROVED** [review]
It looks good @ralfudx, thank you for all the changes 🌟
